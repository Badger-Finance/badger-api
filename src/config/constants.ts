import { ethers } from 'ethers';

const RAW_TOKENS = {
  // eth tokens
  BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
  DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',
  WBTC: '0x2260fac5e5542a773aa44fbcfedf7c193bc2c599',
  WETH: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
  SUSHI: '0x6b3595068778dd592e39a122f4f5a5cf09c90fe2',

  // bsc tokens
  CAKE: '0x0E09FaBB73Bd3Ade0a17ECC321fD13a19e81cE82',
  WBNB: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
  BTCB: '0x7130d2A12B9BCbFAe4f2634d864A1Ee1Ce3Ead9c',
  BSC_BADGER: '0x753fbc5800a8C8e3Fb6DC6415810d627A387Dfc9',
  BSC_BBADGER: '0x1F7216fdB338247512Ec99715587bb97BBf96eae',
  BSC_BDIGG: '0x5986D5c77c65e5801a5cAa4fAE80089f870A71dA',

  // curve tokens
  CRV_RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
  CRV_TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
  CRV_SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',

  // uniswap tokens
  UNI_BADGER_WBTC: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
  UNI_DIGG_WBTC: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

  // sushiswap tokens
  SUSHI_ETH_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
  SUSHI_BADGER_WBTC: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
  SUSHI_DIGG_WBTC: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',

  // pancakeswap tokens
  PANCAKE_BNB_BTCB: '0x356b7d0d3c54F22C82B7a670C6Ba9E2381b0624c',
  PANCAKE_BBADGER_BTCB: '0x87Ae7b5c43D4e160cDB9427a78BA87B9503ee37b',
  PANCAKE_BDIGG_BTCB: '0xfbfa92e037e37F946c0105902640914E3aCe6752',

  // eth vault tokens
  BBADGER: '0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28',
  BDIGG: '0x7e7e112a68d8d2e221e11047a72ffc1065c38e1a',
  BUNI_BADGER_WBTC: '0x235c9e24d3fb2fafd58a2e49d454fdcd2dbf7ff1',
  BUNI_DIGG_WBTC: '0xc17078fdd324cc473f8175dc5290fae5f2e84714',
  BSUSHI_ETH_WBTC: '0x758a43ee2bff8230eeb784879cdcff4828f2544d',
  BSUSHI_BADGER_WBTC: '0x1862a18181346ebd9edaf800804f89190def24a5',
  BSUSHI_DIGG_WBTC: '0x88128580acdd9c04ce47afce196875747bf2a9f6',
  BCRV_SBTC: '0xd04c48a53c111300ad41190d63681ed3dad998ec',
  BCRV_RENBTC: '0x6def55d2e18486b9ddfaa075bc4e4ee0b28c1545',
  BCRV_TBTC: '0xb9d076fde463dbc9f915e5392f807315bf940334',
  BCRV_HRENBTC: '0xaf5a1decfa95baf63e0084a35c62592b774a2a87',
  BVYWBTC: '0x4b92d19c11435614cd49af1b589001b7c08cd4d5',

  // bsc vault tokens
  BPANCAKE_BNB_BTCB: '0xaf4B9C4b545D5324904bAa15e29796D2E2f90813',
  BPANCAKE_BBADGER_BTCB: '0x857F91f735f4B03b19D2b5c6E476C73DB8241F55',
  BPANCAKE_BDIGG_BTCB: '0xa861Ba302674b08f7F2F24381b705870521DDfed',
};

export const TOKENS = Object.fromEntries(
  Object.entries(RAW_TOKENS).map(([key, val]) => [key, ethers.utils.getAddress(val)]),
);

export enum Protocol {
  Curve = 'curve',
  Sushiswap = 'sushiswap',
  Uniswap = 'uniswap',
  Pancakeswap = 'pancakeswap',
  Yearn = 'yearn',
}

// data point constants - index two times per hour, 48 per day
export const CURRENT = 0;
export const ONE_DAY = 24 * 2;
export const ONE_MINUTE_MS = 60 * 1000;
export const THREE_DAYS = ONE_DAY * 3;
export const SEVEN_DAYS = ONE_DAY * 7;
export const THIRTY_DAYS = ONE_DAY * 30;
export const SAMPLE_DAYS = THIRTY_DAYS + 1;
export const ONE_YEAR_MS = 365 * 24 * 60 * 60 * 1000;

// data access constants
export const APY_SNAPSHOTS_DATA = process.env.APY_SNAPSHOTS_DATA || 'MISSING REQUIRED ENV VAR';
export const ASSET_DATA = process.env.ASSET_DATA || 'MISSING REQUIRED ENV VAR';
export const PRICE_DATA = process.env.PRICE_DATA || 'MISSING REQUIRED ENV VAR';
export const SETT_SNAPSHOTS_DATA = process.env.SETT_SNAPSHOTS_DATA || 'MISSING REQUIRED ENV VAR';
export const REWARD_DATA = process.env.REWARD_DATA || 'MISSING REQUIRED ENV VAR';
export const UNISWAP_URL = process.env.UNISWAP || 'MISSING REQUIRED ENV VAR';
export const SUSHISWAP_URL = process.env.SUSHISWAP || 'MISSING REQUIRED ENV VAR';
export const PANCAKESWAP_URL = process.env.PANCAKESWAP || 'MISSING REQUIRED ENV VAR';
export const BADGER_URL = process.env.BADGER || 'MISSING REQUIRED ENV VAR';
export const BADGER_BSC_URL = process.env.BADGER_BSC || 'MISSING REQUIRED ENV VAR';
export const BADGER_DAO_URL = process.env.BADGER_DAO || 'MISSING REQUIRED ENV VAR';
export const MASTERCHEF_URL = process.env.MASTERCHEF || 'MISSING REQUIRED ENV VAR';
export const BOUNCER_PROOFS = process.env.BOUNCER_PROOFS || 'MISSING REQUIRED ENV VAR';

// third party api constants
export const CURVE_API_URL = 'https://stats.curve.fi/raw-stats/apys.json';
export const COINGECKO_URL = 'https://api.coingecko.com/api/v3/simple';

export enum Provider {
  Cloudflare = 'https://cloudflare-eth.com/',
  Binance = 'https://bsc-dataseed.binance.org/',
  MyEtherWallet = 'https://nodes.mewapi.io/rpc/eth',
}
export const SUSHI_CHEF = '0xc2EdaD668740f1aA35E4D8f227fB8E17dcA888Cd';
export const PANCAKE_CHEF = '0x73feaa1eE314F8c655E354234017bE2193C9E24E';
