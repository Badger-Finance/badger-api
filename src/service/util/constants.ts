export const TOKENS = {
	// general tokens
	RENBTC: '0x49849c98ae39fff122806c06791fa73784fb3675',
	BADGER: '0x3472a5a71965499acd81997a54bba8d852c6e53d',
	TBTC: '0x64eda51d3ad40d56b9dfc5554e06f94e1dd786fd',
	SBTC: '0x075b1bb99792c9e1041ba13afef80c91a1e70fb3',
	DIGG: '0x798d1be841a82a273720ce31c822c61a67a601c3',

	// uniswap tokens
	UNI_BADGER: '0xcd7989894bc033581532d2cd88da5db0a4b12859',
	UNI_DIGG: '0xe86204c4eddd2f70ee00ead6805f917671f56c52',

	// sushiswap tokens
	SUSHI_WBTC: '0xceff51756c56ceffca006cd410b03ffc46dd3a58',
	SUSHI_BADGER: '0x110492b31c59716ac47337e616804e3e3adc0b4a',
	SUSHI_DIGG: '0x9a13867048e01c663ce8ce2fe0cdae69ff9f35e3',
};
