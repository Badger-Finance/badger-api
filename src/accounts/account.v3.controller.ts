import { Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, QueryParams } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { AccountsService } from './accounts.service';
import { AccountModel } from './interfaces/account-model.interface';
import { QueryParamError } from '../errors/validation/query.param.error';

@Controller('/account')
export class AccountV3Controller {
  @Inject()
  accountsService!: AccountsService;

  @Get()
  @ContentType('json')
  @Summary('Get badger user account information')
  @Description(
    'Return key user information for a given account. Includes positions, earnings from use, and claimable balances.',
  )
  @Returns(200, AccountModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid account')
  async getAccount(
    @QueryParams('accountId') userId: string,
    @QueryParams('chain') chain: Network,
  ): Promise<AccountModel> {
    if (!userId) throw new QueryParamError('userId');

    return this.accountsService.getAccount(Chain.getChain(chain), userId);
  }
}
