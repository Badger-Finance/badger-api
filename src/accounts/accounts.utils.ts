import { Network } from '@badger-dao/sdk';
import { ethers } from 'ethers';
import { GraphQLClient } from 'graphql-request';
import { getDataMapper } from '../aws/dynamodb.utils';
import { Chain } from '../chains/config/chain.config';
import { TOKENS } from '../config/tokens.config';
import {
  getSdk,
  OrderDirection,
  User_OrderBy,
  UserQuery,
  UserSettBalance,
  UsersQuery,
} from '../graphql/generated/badger';
import { LeaderBoardType } from '../leaderboards/enums/leaderboard-type.enum';
import { CachedBoost } from '../leaderboards/interface/cached-boost.interface';
import { getPrice, inCurrency } from '../prices/prices.utils';
import { getCachedSett, getSettDefinition } from '../setts/setts.utils';
import { cachedTokenBalanceToTokenBalance, formatBalance, getSettTokens, getToken } from '../tokens/tokens.utils';
import { Account } from './interfaces/account.interface';
import { CachedAccount } from './interfaces/cached-account.interface';
import { CachedSettBalance } from './interfaces/cached-sett-balance.interface';

export function defaultBoost(address: string): CachedBoost {
  return {
    leaderboard: LeaderBoardType.BadgerBoost,
    rank: 0,
    address,
    boost: 1,
    stakeRatio: 0,
    nftMultiplier: 0,
    nativeBalance: 0,
    nonNativeBalance: 0,
  };
}

export async function getUserAccount(chain: Chain, accountId: string): Promise<UserQuery> {
  const badgerGraphqlClient = new GraphQLClient(chain.graphUrl);
  const badgerGraphqlSdk = getSdk(badgerGraphqlClient);
  return badgerGraphqlSdk.User({
    id: accountId.toLowerCase(),
    orderDirection: OrderDirection.Asc,
  });
}

export async function getUserAccounts(chain: Chain, accounts: string[]): Promise<UsersQuery> {
  const badgerGraphqlClient = new GraphQLClient(chain.graphUrl);
  const badgerGraphqlSdk = getSdk(badgerGraphqlClient);
  return badgerGraphqlSdk.Users({
    where: {
      id_in: accounts.map((acc) => acc.toLowerCase()),
    },
  });
}

export async function getAccounts(chain: Chain): Promise<string[]> {
  const badgerGraphqlClient = new GraphQLClient(chain.graphUrl);
  const badgerGraphqlSdk = getSdk(badgerGraphqlClient);

  const accounts: string[] = [];

  let lastAddress: string | undefined;
  const pageSize = 1000;
  while (true) {
    try {
      const userPage = await badgerGraphqlSdk.Users({
        first: pageSize,
        where: { id_gt: lastAddress },
        orderBy: User_OrderBy.Id,
        orderDirection: OrderDirection.Asc,
      });
      if (!userPage || !userPage.users || userPage.users.length === 0) {
        break;
      }
      const { users } = userPage;
      lastAddress = users[users.length - 1].id;
      users.forEach((user) => accounts.push(ethers.utils.getAddress(user.id)));
    } catch (err) {
      break;
    }
  }

  return accounts;
}

export function cachedAccountToAccount(cachedAccount: CachedAccount, network?: Network): Account {
  const balances = cachedAccount.balances
    .filter((bal) => !network || bal.network === network)
    .map((bal) => ({
      ...bal,
      tokens: bal.tokens.map((token) => cachedTokenBalanceToTokenBalance(token)),
      earnedTokens: bal.earnedTokens.map((token) => cachedTokenBalanceToTokenBalance(token)),
    }));
  const multipliers = Object.fromEntries(cachedAccount.multipliers.map((entry) => [entry.address, entry.multiplier]));
  const claimableBalances = cachedAccount.claimableBalances.filter((bal) => !network || bal.network === network);
  const claimableBalancesMap = Object.fromEntries(claimableBalances.map((bal) => [bal.address, bal.balance]));
  const account: Account = {
    ...cachedAccount,
    multipliers,
    balances,
    claimableBalances,
    claimableBalancesMap,
  };
  delete account.multipliers[TOKENS.BICVX];
  return account;
}

export async function getCachedAccount(address: string): Promise<CachedAccount> {
  const checksummedAccount = ethers.utils.getAddress(address);
  const defaultAccount = {
    address: checksummedAccount,
    boost: 0,
    boostRank: 0,
    multipliers: [],
    value: 0,
    earnedValue: 0,
    balances: [],
    claimableBalances: [],
    nativeBalance: 0,
    nonNativeBalance: 0,
  };
  try {
    const mapper = getDataMapper();
    for await (const item of mapper.query(
      CachedAccount,
      { address: checksummedAccount },
      { limit: 1, scanIndexForward: false },
    )) {
      return {
        ...defaultAccount,
        ...item,
      };
    }
    return defaultAccount;
  } catch (err) {
    return defaultAccount;
  }
}

export async function toSettBalance(
  chain: Chain,
  settBalance: UserSettBalance,
  currency?: string,
): Promise<CachedSettBalance> {
  const settDefinition = getSettDefinition(chain, settBalance.sett.id);
  const { netShareDeposit, grossDeposit, grossWithdraw } = settBalance;
  const { ppfs } = await getCachedSett(settDefinition);

  const depositToken = getToken(settDefinition.depositToken);
  const settToken = getToken(settDefinition.settToken);
  const currentTokens = formatBalance(netShareDeposit, settToken.decimals);
  let depositTokenDecimals = depositToken.decimals;
  if (depositToken.address === TOKENS.DIGG) {
    depositTokenDecimals = settToken.decimals;
  }
  const depositedTokens = formatBalance(grossDeposit, depositTokenDecimals);
  const withdrawnTokens = formatBalance(grossWithdraw, depositTokenDecimals);
  const balanceTokens = currentTokens * ppfs;
  const earnedBalance = balanceTokens - depositedTokens + withdrawnTokens;
  const [depositTokenPrice, earnedTokens, tokens] = await Promise.all([
    getPrice(settDefinition.depositToken),
    getSettTokens(settDefinition, earnedBalance, currency),
    getSettTokens(settDefinition, balanceTokens, currency),
  ]);

  return Object.assign(new CachedSettBalance(), {
    network: chain.network,
    id: settDefinition.settToken,
    name: settDefinition.name,
    asset: depositToken.symbol,
    ppfs,
    balance: balanceTokens,
    value: inCurrency(depositTokenPrice, currency) * balanceTokens,
    tokens,
    earnedBalance: earnedBalance,
    earnedValue: inCurrency(depositTokenPrice, currency) * earnedBalance,
    earnedTokens,
    depositedBalance: depositedTokens,
    withdrawnBalance: withdrawnTokens,
  });
}

export async function getCachedBoost(address: string): Promise<CachedBoost> {
  const mapper = getDataMapper();
  for await (const entry of mapper.query(
    CachedBoost,
    { address: ethers.utils.getAddress(address) },
    { limit: 1, indexName: 'IndexLeaderBoardRankOnAddress' },
  )) {
    return entry;
  }
  return defaultBoost(address);
}
