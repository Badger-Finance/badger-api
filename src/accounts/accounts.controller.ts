import { Controller, Get, Inject, PathParams, QueryParams } from '@tsed/common';
import { ContentType, Description, Hidden, Returns, Summary } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { ChainNetwork } from '../chains/enums/chain-network.enum';
import { AccountsService } from './accounts.service';
import { AccountModel } from './interfaces/account-model.interface';
import { UnclaimedRewards } from './interfaces/unclaimed-rewards.interface';

@Controller('/accounts')
export class AccountsController {
  @Inject()
  accountsService!: AccountsService;

  @Hidden()
  @Get('/allClaimable')
  @ContentType('json')
  async getAllBadgerTreeRewards(@QueryParams('page') page?: number): Promise<UnclaimedRewards> {
    return this.accountsService.getAllUnclaimed(page ?? 1);
  }

  @Get('/:accountId')
  @ContentType('json')
  @Summary('Get badger user account information')
  @Description(
    'Return key user information for a given account. Includes positions, earnings from use, and claimable balances.',
  )
  @Returns(200, AccountModel)
  @(Returns(400).Description('Not a valid chain'))
  @(Returns(404).Description('Not a valid account'))
  async getAccount(
    @PathParams('accountId') userId: string,
    @QueryParams('chain') chain?: ChainNetwork,
  ): Promise<AccountModel> {
    return this.accountsService.getAccount(Chain.getChain(chain), userId);
  }
}
