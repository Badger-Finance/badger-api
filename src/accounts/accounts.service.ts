import { Account } from '@badger-dao/sdk';
import { Service } from '@tsed/di';
import { ethers } from 'ethers';

import { Chain } from '../chains/config/chain.config';
import { InvalidAddrError } from '../errors/validation/invalid.addr.error';
import { getCachedAccount } from './accounts.utils';

@Service()
export class AccountsService {
  async getAccount(chain: Chain, address: string): Promise<Account> {
    let checksumAddress = address;
    try {
      checksumAddress = ethers.utils.getAddress(address);
    } catch {
      throw new InvalidAddrError(`${address}`);
    }
    return getCachedAccount(chain, checksumAddress);
  }
}
