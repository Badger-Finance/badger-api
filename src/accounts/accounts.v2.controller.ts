import { Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, PathParams, QueryParams } from '@tsed/common';
import { ContentType, Deprecated, Description, Returns, Summary } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { AccountsService } from './accounts.service';
import { AccountModel } from './interfaces/account-model.interface';

@Deprecated()
@Controller('/accounts')
export class AccountsV2Controller {
  @Inject()
  accountsService!: AccountsService;

  @Get('/:accountId')
  @ContentType('json')
  @Summary('Get badger user account information')
  @Description(
    'Return key user information for a given account. Includes positions, earnings from use, and claimable balances.',
  )
  @Returns(200, AccountModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid account')
  async getAccount(
    @PathParams('accountId') userId: string,
    @QueryParams('chain') chain: Network,
  ): Promise<AccountModel> {
    return this.accountsService.getAccount(Chain.getChain(chain), userId);
  }
}
