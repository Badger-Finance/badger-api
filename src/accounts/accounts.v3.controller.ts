import { Network } from "@badger-dao/sdk";
import { Controller, Get, Inject, QueryParams } from "@tsed/common";
import { ContentType, Description, Returns, Summary } from "@tsed/schema";

import { Chain } from "../chains/config/chain.config";
import { QueryParamError } from "../errors/validation/query.param.error";
import { AccountsService } from "./accounts.service";
import { AccountModel } from "./interfaces/account-model.interface";

@Controller("/accounts")
export class AccountsV3Controller {
  @Inject()
  accountsService!: AccountsService;

  @Get()
  @ContentType("json")
  @Summary("Get badger user account information")
  @Description(
    "Return key user information for a given account. Includes positions, earnings from use, and claimable balances."
  )
  @Returns(200, AccountModel)
  @Returns(400).Description("Not a valid chain")
  @Returns(404).Description("Not a valid account")
  async getAccount(
    @QueryParams("address") address: string,
    @QueryParams("chain") chain?: Network
  ): Promise<AccountModel> {
    if (!address) throw new QueryParamError("address");

    return this.accountsService.getAccount(Chain.getChain(chain), address);
  }
}
