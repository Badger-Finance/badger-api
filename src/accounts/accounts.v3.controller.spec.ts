import { PlatformServerless } from "@tsed/platform-serverless";
import { PlatformServerlessTest } from "@tsed/platform-serverless-testing";

import { NetworkStatus } from "../errors/enums/network-status.enum";
import { TEST_ADDR } from "../test/constants";
import { setupMockChain } from "../test/mocks.utils";
import * as accountsUtils from "./accounts.utils";
import { setupMockAccounts } from "./accounts.v2.controller.spec";
import { AccountsV3Controller } from "./accounts.v3.controller";

describe("accounts.v3.controller", () => {
  beforeEach(
    PlatformServerlessTest.bootstrap(PlatformServerless, {
      lambda: [AccountsV3Controller]
    })
  );
  afterEach(() => PlatformServerlessTest.reset());

  beforeEach(() => {
    setupMockChain();
    setupMockAccounts();
  });

  describe("GET /v3/account", () => {
    describe("with no specified account", () => {
      it("returns a not found response", async () => {
        const { body, statusCode } = await PlatformServerlessTest.request.get("/accounts");
        expect(statusCode).toEqual(NetworkStatus.BadRequest);
        expect(JSON.parse(body)).toMatchSnapshot();
      });
    });

    describe("with an invalid account input", () => {
      it("returns a bad request response", async () => {
        const { body, statusCode } = await PlatformServerlessTest.request
          .get("/accounts")
          .query({ address: "0xjintao" });
        expect(statusCode).toEqual(NetworkStatus.BadRequest);
        expect(JSON.parse(body)).toMatchSnapshot();
      });
    });

    describe("with a non participant account", () => {
      it("returns a default account response", async () => {
        jest.spyOn(accountsUtils, "getCachedAccount").mockImplementation(async (_chain, address) => ({
          address,
          value: 0,
          earnedValue: 0,
          boost: 1,
          rank: 1035,
          boostRank: 1035,
          multipliers: {},
          data: {},
          claimableBalances: {},
          stakeRatio: 0,
          nftBalance: 0,
          bveCvxBalance: 0,
          diggBalance: 0,
          nativeBalance: 0,
          nonNativeBalance: 0
        }));
        const { body, statusCode } = await PlatformServerlessTest.request
          .get(`/accounts`)
          .query({ address: TEST_ADDR });
        expect(statusCode).toEqual(NetworkStatus.Success);
        expect(JSON.parse(body)).toMatchSnapshot();
      });
    });

    describe("with a participant account", () => {
      it("returns a cached account response", async () => {
        jest.spyOn(accountsUtils, "getCachedAccount").mockImplementation(async (_chain, address) => ({
          address,
          value: 10,
          earnedValue: 1,
          boost: 2000,
          rank: 1,
          boostRank: 1,
          multipliers: {},
          data: {},
          claimableBalances: {},
          stakeRatio: 1,
          nftBalance: 3,
          bveCvxBalance: 0,
          diggBalance: 0,
          nativeBalance: 3,
          nonNativeBalance: 5
        }));
        const { body, statusCode } = await PlatformServerlessTest.request
          .get(`/accounts`)
          .query({ address: TEST_ADDR });
        expect(statusCode).toEqual(NetworkStatus.Success);
        expect(JSON.parse(body)).toMatchSnapshot();
      });
    });
  });
});
