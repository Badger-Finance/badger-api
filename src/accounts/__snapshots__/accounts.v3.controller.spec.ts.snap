// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accounts.v3.controller GET /v3/account with a non participant account returns a default account response 1`] = `
Object {
  "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "boost": 1,
  "boostRank": 1035,
  "bveCvxBalance": 0,
  "claimableBalances": Object {},
  "data": Object {},
  "diggBalance": 0,
  "earnedValue": 0,
  "nativeBalance": 0,
  "nftBalance": 0,
  "nonNativeBalance": 0,
  "stakeRatio": 0,
  "value": 0,
}
`;

exports[`accounts.v3.controller GET /v3/account with a participant account returns a cached account response 1`] = `
Object {
  "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "boost": 2000,
  "boostRank": 1,
  "bveCvxBalance": 0,
  "claimableBalances": Object {},
  "data": Object {},
  "diggBalance": 0,
  "earnedValue": 1,
  "nativeBalance": 3,
  "nftBalance": 3,
  "nonNativeBalance": 5,
  "stakeRatio": 1,
  "value": 10,
}
`;

exports[`accounts.v3.controller GET /v3/account with an invalid account input returns a bad request response 1`] = `
Object {
  "errors": Array [],
  "message": "Addr 0xjintao is not a valid account",
  "name": "Error",
  "status": 400,
}
`;

exports[`accounts.v3.controller GET /v3/account with no specified account returns a not found response 1`] = `
Object {
  "errors": Array [],
  "message": "Query param address invalid",
  "name": "Error",
  "status": 400,
}
`;
